docs: add documentation for all secure acl introduction work (#5640)
hashicorp/consul: website/source/api/acl/identity-providers.html.md

---
layout: api
page_title: ACL Identity Providers - HTTP API
sidebar_current: api-acl-identity-providers
description: |-
  The /acl/idp endpoints manage Consul's ACL Identity Providers.
---

> -> **1.5.0+:**  The APIs are available in Consul versions 1.5.0 and later. The documentation for the legacy ACL API is [here](/api/acl/legacy.html)

Reviewer: I'm not sure we want to refer people to legacy docs, since they could be using the new system just not the latest. This seems like a group discussion.

Reviewer: This will apply to all three new API docs

Author: Maybe because these concepts are completely new and you are far less likely to end up on these pages by accident I could omit the second sentence in these warnings?

Reviewer: I think its ok to omit the second sentence. I would hope by this point in the documentation the users are aware there are two different ACL systems and are being careful. We can revisit if users start having issues here.

The `/acl/idp` endpoints [create](#create-an-identity-provider),
[read](#read-an-identity-provider), [update](#update-an-identity-provider),
[list](#list-identity-providers) and [delete](#delete-an-identity-provider)
ACL identity providers in Consul.  For more information about ACLs, please see
the [ACL Guide](/docs/guides/acl.html).

## Create an Identity Provider

This endpoint creates a new ACL identity provider.

| Method | Path                         | Produces                   |
| ------ | ---------------------------- | -------------------------- |
| `PUT`  | `/acl/idp`                   | `application/json`         |

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and
[required ACLs](/api/index.html#acls).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `NO`             | `none`            | `none`        | `acl:write`  |

### Parameters

- `Name` `(string: <required>)` - Specifies a name for the ACL identity

>   provider. The name can only contain alphanumeric characters as well as `-`

Reviewer: ```suggestion
  provider. The name can only contain alphanumeric characters, dashes `-`,
```

>   and `_` and must be unique. This field is immutable.

Reviewer: ```suggestion
 underscores `_`, and must be unique. This field is immutable.
```

    "KubernetesServiceAccountJWT": "eyJhbGciOiJSUzI1NiIsImtpZCI6IiJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6ImNvbnN1bC1pZHAtdG9rZW4tcmV2aWV3LWFjY291bnQtdG9rZW4tbHBteHYiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoiY29uc3VsLWlkcC10b2tlbi1yZXZpZXctYWNjb3VudCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6IjZiMGU4NzkxLTU3ZDItMTFlOS1iYzJhLTQ4ZTZjOGI4ZWNiNSIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpkZWZhdWx0OmNvbnN1bC1pZHAtdG9rZW4tcmV2aWV3LWFjY291bnQifQ.bdhKU-r_5jJeg9kKIhqmZDxqSN4rD0wSvpDc_JBfP8WqLeaA7bZctt9oQW8FtNqMtlnM7eLsRkHspqNuudXADz5FWuWmU59jqccfrwazYm5qHpOWdYcVTZKeH1cCxheWW0GDuTPrXINcJoRFS6wxhytBLC-aBy3pObGjmEpApr3G9SAcY9cT5RgkglTx9yJhQHxIICHg9ktxPKy86WkaJq4k8L6lK-3oK0Ul6gy4Dgmx_DvgBrN_dAOmWWLX3nYUT62TpWLDto6wmXW9Z354ziv_XBlmq52K5lQw8pq1B7M8scUanSJ693CGippMbutjV3RJ08UkfgsN1DHWVUQ4pQ"
}
```

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

    "CreateIndex": 15,
    "ModifyIndex": 15
}
```

## Read an Identity Provider

This endpoint reads an ACL identity provider with the given name. If no
identity provider exists with the given name, a 404 is returned instead of a
200 response.

| Method | Path                         | Produces                   |
| ------ | ---------------------------- | -------------------------- |
| `GET`  | `/acl/idp/:name`             | `application/json`         |

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and
[required ACLs](/api/index.html#acls).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `YES`            | `all`             | `none`        | `acl:read`   |

### Parameters

- `name` `(string: <required>)` - Specifies the name of the ACL identity
  provider to read. This is required and is specified as part of the URL path.

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

    "KubernetesServiceAccountJWT": "eyJhbGciOiJSUzI1NiIsImtpZCI6IiJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6ImNvbnN1bC1pZHAtdG9rZW4tcmV2aWV3LWFjY291bnQtdG9rZW4tbHBteHYiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoiY29uc3VsLWlkcC10b2tlbi1yZXZpZXctYWNjb3VudCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6IjZiMGU4NzkxLTU3ZDItMTFlOS1iYzJhLTQ4ZTZjOGI4ZWNiNSIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpkZWZhdWx0OmNvbnN1bC1pZHAtdG9rZW4tcmV2aWV3LWFjY291bnQifQ.bdhKU-r_5jJeg9kKIhqmZDxqSN4rD0wSvpDc_JBfP8WqLeaA7bZctt9oQW8FtNqMtlnM7eLsRkHspqNuudXADz5FWuWmU59jqccfrwazYm5qHpOWdYcVTZKeH1cCxheWW0GDuTPrXINcJoRFS6wxhytBLC-aBy3pObGjmEpApr3G9SAcY9cT5RgkglTx9yJhQHxIICHg9ktxPKy86WkaJq4k8L6lK-3oK0Ul6gy4Dgmx_DvgBrN_dAOmWWLX3nYUT62TpWLDto6wmXW9Z354ziv_XBlmq52K5lQw8pq1B7M8scUanSJ693CGippMbutjV3RJ08UkfgsN1DHWVUQ4pQ"
}
```

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

    "Type": "kubernetes",
    "KubernetesHost": "https://192.0.2.42:8443",
    "KubernetesCACert": "-----BEGIN CERTIFICATE-----\nMIIC5zCCAc+gAwIBAgIBATANBgkqhkiG9w0BAQsFADAVMRMwEQYDVQQDEwptaW5p\na3ViZUNBMB4XDTE5MDEwODE4NTAwNFoXDTI5MDEwNjE4NTAwNFowFTETMBEGA1UE\nAxMKbWluaWt1YmVDQTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALna\nHbH7BirRjzbExPgGdsikbPUtwmUqyGvJ5SGGho0RkkL/HIL5D154CD8CvAMEbSMt\n20FBqbI5h5JeJ9gYUdsPLwnMVI2r2yvJuzuPxtxiyf63DU7XxXi3v6QP6J0qRhDr\nlEhp5K2Xrd7qIxABZokNqPF8o5cfSjVgoREn6wKpmcCsf7rAqRpvMZWWYwsyAZJK\nEqWASwY6PmcwN3tboLoPBv7ZSJ8Q3y6RBypeNbngJgylNXP/OGwNO9Nm/HpmL1vc\nlR92K+tJE43XONkjkuXicC+ImhiX8cA6elmNIPog0UcpXal8CrDRug9MiGV3AJit\npxDjRFa163mjx536bE0CAwEAAaNCMEAwDgYDVR0PAQH/BAQDAgKkMB0GA1UdJQQW\nMBQGCCsGAQUFBwMCBggrBgEFBQcDATAPBgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3\nDQEBCwUAA4IBAQB+MRDedyRpsL08sDuU8EnJ/tEPZ6B7uTTBFrPyLoxbc7DfZEm0\nPlwyMk04TNlxoWLlcs4DBI/goVpER9kVoSd9CWmwG62Mqptk/AsyBlxqh0ZbHkAU\ndPHBXHFdoCkGnM//HDc8DW9zKGjv2wlKptPUxq1tkAX5aKUktsOVpmKH0YgKLuWA\n3Hvhgh2Gd/ssIwp1pHbWHKZ9+HS1etwHqgSneLpQ50K4iv0Rk7yxQmfCQC2wiz4q\n5T4dUXB28k0hl/Lx1QCfMvy20btyhnPleeW0xiQZ4Yq/XYv11Ez81duOKjZGYick\nZvw0ZvX68ssuWgbrWHkFSFJyBskBZOl1Rtln\n-----END CERTIFICATE-----\n",

>     "KubernetesServiceAccountJWT": "eyJhbGciOiJSUzI1NiIsImtpZCI6IiJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6ImNvbnN1bC1pZHAtdG9rZW4tcmV2aWV3LWFjY291bnQtdG9rZW4tbHBteHYiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoiY29uc3VsLWlkcC10b2tlbi1yZXZpZXctYWNjb3VudCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6IjZiMGU4NzkxLTU3ZDItMTFlOS1iYzJhLTQ4ZTZjOGI4ZWNiNSIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpkZWZhdWx0OmNvbnN1bC1pZHAtdG9rZW4tcmV2aWV3LWFjY291bnQifQ.bdhKU-r_5jJeg9kKIhqmZDxqSN4rD0wSvpDc_JBfP8WqLeaA7bZctt9oQW8FtNqMtlnM7eLsRkHspqNuudXADz5FWuWmU59jqccfrwazYm5qHpOWdYcVTZKeH1cCxheWW0GDuTPrXINcJoRFS6wxhytBLC-aBy3pObGjmEpApr3G9SAcY9cT5RgkglTx9yJhQHxIICHg9ktxPKy86WkaJq4k8L6lK-3oK0Ul6gy4Dgmx_DvgBrN_dAOmWWLX3nYUT62TpWLDto6wmXW9Z354ziv_XBlmq52K5lQw8pq1B7M8scUanSJ693CGippMbutjV3RJ08UkfgsN1DHWVUQ4pQ",

Reviewer: I think we should truncate all these tokens and certificates. If necessary we can call that out at the beginning of the doc.

Author: I'll just do it how I noticed the vault docs do it: https://www.vaultproject.io/api/auth/kubernetes/index.html

[role binding rules](/api/acl/role-binding-rules.html) as well as marking any
outstanding [tokens](/api/acl/tokens.html) created from this identity provider
as eligible for deletion.

| Method   | Path                      | Produces                   |
| -------- | ------------------------- | -------------------------- |
| `DELETE` | `/acl/idp/:name`          | `application/json`         |

Even though the return type is application/json, the value is either true or
false indicating whether the delete succeeded.

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and
[required ACLs](/api/index.html#acls).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `NO`             | `none`            | `none`        | `acl:write`  |

### Parameters

- `name` `(string: <required>)` - Specifies the name of the ACL identity
  provider to delete. This is required and is specified as part of the URL
  path.

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

$ curl -X DELETE \
    http://127.0.0.1:8500/v1/acl/idp/minikube
```

### Sample Response
```json
true
```

## List Identity Providers

This endpoint lists all the ACL identity providers.

| Method | Path                         | Produces                   |
| ------ | ---------------------------- | -------------------------- |
| `GET`  | `/acl/idps`                 | `application/json`         |

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and
[required ACLs](/api/index.html#acls).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `YES`            | `all`             | `none`        | `acl:read`   |

## Sample Request

> ```text

Reviewer: ```suggestion
```sh
```
